### PR TITLE
[WIP] Adds support for installing Allihoopa in drop flow

### PR DESCRIPTION
**Pull request by JagCesar**
After a piece has been dropped the user can choose to view the piece.
The SDK now tried to open the piece in the Allihoopa app, if it's
installed. If it isn't installed on the current device the user is shown
a modal view where she can decide to install the app or open the piece
in Safari.

- [x] Design fixes
- [x] Japanese translation
- [x] Update Cartfile with correct release of AllihoopaCore

### CODE REVIEW

**Comment by JagCesar** [comment]
simulator screen shot 1 jun 2017 17 44 19
simulator screen shot 1 jun 2017 17 44 13

**Comment by Boerworz** [comment]
Comparing the screenshots to the mockups in Zeplin it seems that the top margin of the Allihoopa app icon is less than specified, and the "Meet like-minded music doers […]" text should use `NSTextAlignment.center`.

**Comment by JagCesar** [comment]
You are correct señor, forgot to onion skin the design. 🙈

File: Allihoopa/Drop/DropDoneViewController.m [hunk 1]
```diff
@@ -61,14 +62,19 @@ - (void)viewDidLayoutSubviews {
...
-	_playerURL = [url copy];
 	_coverImage = coverImage;
+    [self setPieceIdentifier:identifier];
 }
 
 - (IBAction)viewOnAllihoopa {
-	[[UIApplication sharedApplication] openURL:_playerURL];
+    NSURL *pieceURL = [[NSURL alloc] initWithString:[NSString stringWithFormat:@"ah-allihoopa://allihoopa.com/s/%@", _pieceIdentifier]];
+    if ([[UIApplication sharedApplication] openURL:pieceURL] == NO) {
+        AllihoopaInstallerViewController *allihoopaInstallerViewController = [[AllihoopaInstallerViewController alloc] initWithPieceIdentifier:_pieceIdentifier NibName:@"AllihoopaInstallerView" bundle:[NSBundle bundleForClass:[AllihoopaInstallerViewController class]]];
```
Comment (Boerworz) [inline Allihoopa/Drop/DropDoneViewController.m]: If we rename `AllihoopaInstallerView.xib` to `AllihoopaInstallerViewController.xib` I believe we don't have to specify the nib name and bundle. Or are the rules different for frameworks?

File: Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m [hunk 2]
```diff
@@ -0,0 +1,136 @@
+#import "AllihoopaInstallerViewController.h"
+@import StoreKit;
```
Comment (Boerworz) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: We've made a choice to support building the SDK with module support disabled (see be8e861). Please replace this with the `#import <StoreKit/StoreKit.h>` syntax.

File: Allihoopa.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -538,6 +561,7 @@
 		FA5FD39F1D90180F0010E978 /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
+				CLANG_ENABLE_MODULES = YES;
```
Comment (Boerworz) [inline Allihoopa.xcodeproj/project.pbxproj]: Please revert these changes if you didn't consciously make them.

File: Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+@property (strong, nonatomic) IBOutlet UIImageView *shadowImageView;
+@property (strong, nonatomic) IBOutlet UIButton *installButton;
+@property (strong, nonatomic) IBOutlet UIButton *viewInBrowserButton;
+
+@end
+
+@implementation AllihoopaInstallerViewController
+
+- (instancetype)initWithPieceIdentifier:(NSString *)pieceIdentifier NibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
+    [self setPieceIdentifier:pieceIdentifier];
```
Comment (Boerworz) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: The superclass' initializer should be invoked before initializing instance variables ([reference](https://developer.apple.com/library/content/documentation/General/Conceptual/CocoaEncyclopedia/Initialization/Initialization.html#//apple_ref/doc/uid/TP40010810-CH6-SW58)):

```objc
if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
    [self setPieceIdentifier:pieceIdentifier];
}
return self;
```

File: Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+- (void)setColors {
+    [_iconImageView setTintColor:[AllihoopaInstallerColors smokehouseGray]];
+    [_titleLabel setTextColor:[AllihoopaInstallerColors birkenstockGray]];
+}
+
+@end
+
+@interface AllihoopaInstallerViewController () <SKStoreProductViewControllerDelegate>
+
+@property (strong, nonatomic) NSString *pieceIdentifier;
```
Comment (Boerworz) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: It's good practice to use `copy` for properties of types that have mutable subclasses to prevent situations where the original value is accidentally mutated after assignment.

File: Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    return [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
+}
+
+- (void)dealloc {
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appWillResignActive) name:UIApplicationDidEnterBackgroundNotification object:nil];
```
Comment (JagCesar) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: Read this line one more time :D

File: Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    [_viewInBrowserButton setTintColor:[AllihoopaInstallerColors adamantiumGray]];
+}
+
+- (void)appWillResignActive {
+    [self dismiss];
+}
+
+- (IBAction)dismiss {
+    if ([self presentedViewController]) {
+        [[self presentedViewController] dismissViewControllerAnimated:true completion:nil];
```
Comment (JagCesar) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: Gah, too much swift in my brain :P

File: Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m (hunk 6 above)
Comment (Boerworz) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: Oops! Then I suggest renaming `-appWillResignActive` :D

File: Allihoopa/Drop/DropDoneViewController.m (hunk 1 above)
Comment (JagCesar) [inline Allihoopa/Drop/DropDoneViewController.m]: Didn't work, I think it's because it's looking for AllihoopaInstallerViewController.xib in the current bundle (SDKExample bundle). I'll give it a second try.

**Comment by JagCesar** [comment]
We need to update the Cartfile as soon as https://github.com/allihoopa/AllihoopaCore-ObjC/pull/6 is merged

**Comment by JagCesar** [comment]
@Boerworz Can you ping me when a new release has been made with https://github.com/allihoopa/AllihoopaCore-ObjC/pull/6?

**Comment by JagCesar** [comment]
This is blocked until https://github.com/allihoopa/Allihoopa-iOS/pull/15/files has been merged

**Comment by Boerworz** [comment]
@JagCesar: The Cartfile can now be updated to point to [AllihoopaCore 2.0.0](https://github.com/allihoopa/AllihoopaCore-ObjC/releases/tag/2.0.0).

**Comment by Boerworz** [comment]
@JagCesar: Since #16 (which updates the Cartfile) has been merged, it's no longer necessary for this pull request to update the Cartfile. It should suffice to rebase this feature branch on `develop`.

**Comment by JagCesar** [comment]
@Boerworz This is now ready to be merged. Please squash the commits :)

File: Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m [hunk 8]
```diff
@@ -87,6 +89,11 @@ - (void)viewDidLayoutSubviews {
     [super viewDidLayoutSubviews];
     [[self scrollView] setScrollEnabled:[self isScrollViewContentSizeHigherThanScrollViewFrame]];
     [[self shadowImageView] setHidden:![self isScrollViewContentSizeHigherThanScrollViewFrame]];
+
+	if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
```
Comment (Boerworz) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: This code shouldn't be in `-viewDidLayoutSubviews`. Please move it to somewhere it will be run only once, e.g. `-viewDidLoad`.

Comment (JagCesar) [inline Allihoopa/AllihoopaInstaller/AllihoopaInstallerViewController.m]: Tried to move it to viewDidLoad, but since we're doing vary traits in the storyboard file to adjust constraint constants it doesn't work.

We either have to do all constraint constant magic in code, or have this if in `viewDidLayoutSubviews`.
